packages/integration-sdk-runtime/src/synchronization/index.ts

```diff
@@ -426,6 +437,94 @@ export async function uploadData<T extends UploadDataLookup, K extends keyof T>(
   );
 }
 
+// Interface for storing both the key value and total size of a given array entry
+interface keyAndSize {
+  key: string;
+  size: number;
+}
+
+// Interface for shrink run results
+interface shrinkRawDataResults {
+  initialSize: number;
+  totalSize: number;
+  itemsRemoved: number;
+  totalTime: number;
+}
+
+/**
+ * Helper function to find the largest entry in an array and return its key
+ * and approximate byte size.  We JSON.stringify as a method to try and have
+ * an apples to apples comparison no matter what the data type of the value is.
+ *
+ * @param data
+ * @returns
+ */
+function findLargestItemKeyAndByteSize(data: any[]): keyAndSize {
+  const largestItem: keyAndSize = { key: '', size: 0 };
+  for (const item in data) {
+    const length = data[item]
+      ? Buffer.byteLength(JSON.stringify(data[item]))
+      : 0;
+    if (length > largestItem.size) {
+      largestItem.key = item;
+      largestItem.size = length;
+    }
+  }
+
+  return largestItem;
+}
+
+/**
+ * Removes data from the rawData of the largest entity until the overall size
+ * of the data object is less than maxSize (defaulted to UPLOAD_SIZE_MAX).
+ *
+ * @param data
+ */
+export function shrinkRawData(
+  data: UploadDataLookup[keyof UploadDataLookup][],
+  maxSize = UPLOAD_SIZE_MAX,
+): shrinkRawDataResults {
+  const startTimeInMilliseconds = Date.now();
+  let totalSize = Buffer.byteLength(JSON.stringify(data));
+  const initialSize = totalSize;
+  let itemsRemoved = 0;
+  const sizeOfTruncated = Buffer.byteLength("'TRUNCATED'");
+
+  while (totalSize > maxSize) {
+    let largestEntity: keyAndSize = { key: '', size: 0 };
+    let largestRawDataEntry: keyAndSize = { key: '', size: 0 };
+    let largestItem: keyAndSize = { key: '', size: 0 };
+
+    // Find largest Entity
+    largestEntity = findLargestItemKeyAndByteSize(data);
+
+    // Find largest _rawData entry (typically 0, but check to be certain)
+    largestRawDataEntry = findLargestItemKeyAndByteSize(
+      data[largestEntity.key]['_rawData'],
```

Comment:
Do we need to access this property in this way anymore? Are we able to true TypeScript typings to just use:

```ts
data[largestEntity.key]._rawData
```

Reply:
We should try to avoid using `any` if possible.

Reply:
Updated in latest commit.